Revision of random utility

ptr-yudai opened the pull request:
<!-- Set a title that summarizes the PR changes. -->

## Type of PR
<!-- This project uses issue-driven development, so please take the appropriate action for each PR type. -->

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue
<!-- If this PR is a PR type A/B-1, please provide a link to the corresponding issue. -->
<!-- If this PR is a PR type B-1, please write "N/A" -->
https://github.com/fuzzuf/fuzzuf/issues/14

## Importance of PR
<!-- Please describe the importance of the PR in terms of the following aspects. -->

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview
<!-- Please provide a summary of this PR. -->
<!-- If this PR is a PR type A/B-1, this PR will be considered as an item in the checklist for the related issue. Please provide a link to the issue comment that contains the checklist. -->
<!-- If this PR is a PR type B-2, unnecessary to reference the issue. Please provide a summary. -->
This PR adds `WalkerDiscreteDistribution` class as mentioned in https://github.com/fuzzuf/fuzzuf/issues/14
In addition, the existing random function is revised so that it can generate values with arbitrary types such as `int`, `size_t`, or `double`.
This PR also adds a function to pick up an element of a vector randomly.

## Concerns (Optional)
<!-- If you have any concerns, please describe them clearly by filling in the relevant checklist items below. If there is anything else you would like to share with the reviewer, please include it. -->

- [x] Performance
- [x] Source Code Quality

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [x] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [x] CI is green or confirmed test run results.
- [x] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

ptr-yudai committed:
:sparkles: Add WalkerDiscreteDistribution

ptr-yudai committed:
:lipstick: Add description to sampler

potetisensei commented on test/util/random.cpp:
```diff
@@ -0,0 +1,137 @@
+
+  size_t iter = 1000000; // large enough
+  for (size_t i = 0; i < iter; i++) {
+    res[s()]++;
```
It would be nicer if this part has an additional check about whether s() is within [0, len).

potetisensei commented on test/util/random.cpp:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  for (size_t i = 0; i < len; i++) {
+    BOOST_CHECK(base[i] * 0.95 < res_p[i] && base[i] * 1.05 > res_p[i]);
```
これ統計的に根拠があるならコメントに一応書いておいてほしく、ないなら根拠のある信頼区間に変えたほうがいいかもしれない

potetisensei replied:
Does this inequation have some statistical basis? Then please comment it.
Otherwise, probably we should consider properly changing these parameters(0.95 and 1.05) by calculating confidence interval.

potetisensei commented on test/util/random.cpp:
```diff
@@ -0,0 +1,137 @@
+  WalkerDiscreteDistribution<double> s1(w1);
+  std::vector<size_t> res1(w1.size());
+  for (size_t i = 0; i < iter; i++) {
+    res1[s1()]++;
```
The same as TestWalkerDiscreteDistributionSimple.

potetisensei commented on test/util/random.cpp:
```diff
@@ -0,0 +1,137 @@
+  for (size_t i = 0; i < w1.size(); i++) {
+    double pa = w1[i] / sum;
+    double ps = static_cast<double>(res1[i]) / static_cast<double>(iter);
+    BOOST_CHECK(pa*0.9 <= ps && ps <= pa*1.1);
```
The same as TestWalkerDiscreteDistributionSimple.

potetisensei reviewed (changes requested):
Overall, this change is great.
I left several comments, so please check them.
Also, it would be appreciated if you have time and replace `std::discrete_distribution` in fuzzuf with this new function.

> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [x] The reviewer assigned more reviewers if needed.
- [x] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [x] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

retrage committed:
build: Add containerized build support

This commit adds Dockerfile and a helper scirpt for containerized build.
It also adds a new GitHub Actions target that publishes the container
image on GitHub Packages.

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>
Co-authored-by: M6M <minhok@ricsec.co.jp>

ptr-yudai committed:
:lipstick: Use hypothesis testing for random test

ptr-yudai committed:
:sparkles: Use WalkerDiscreteDistribution instead of std::discrete_distribution

ptr-yudai commented:
Changed the following codes:

- Change `std::discrete_distribution` to `WalkerDiscreteDistribution`
- Add a constructor which accepts iterators (for `std::array`)
- Add test for `std::array`
- Change the test to verify hit counts

Here is my statistical explanation for the new test:

Let `l` be the size of array, `w = {w_0, w_1, ..., w_(l-1)}` be the array of weights, and `n` be the sum of weights.
If we sample from the discrete distribution `M` times, we can define our hypothesis as the following:

- Null hypothesis `H_0`: `i` appears `w_i * n / M` times out of the `M` samples
- Alternative hypothesis `H_1`: `i` does not appear `w_i * n / M` times out of the `M` samples

On selecting the number, the sampling results in eather of the 2 patterns: getting `i` or not.
Therefore, the probability `i` appears `X` times out of the `M` samples according to a binomial distribution `Bi(M, w_i / n)`.
Let the significance level `alpha` be 0.05 if `M` is large enough, 0.01 if `M` is not large enough, respectively.
By putting `p = w_i / n`, the statistic `z` is then denoted by `z = (X - Mp) / sqrt(Mp(1-p))` and this follows a standard normal distribution `N(0, 1)`.
As we put `alpha` be 0.05 or 0.01, we need to check if `|z|` is smaller than Z_(0.025)=1.96 or Z_(0.005)=2.58, respectively.

retrage committed:
docs: Add docs/building.md and references

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>
Co-authored-by: M6M <minhok@ricsec.co.jp>

retrage committed:
docs: Move TUTORIAL.md under docs

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

retrage committed:
docs: Rename TUTORIAL.md to tutorial.md

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

retrage committed:
build: Add devcontainer.json

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

retrage committed:
github: Fix container image tag

retrage committed:
scripts: Fix container image tag

retrage committed:
docs: Add note on release build with container

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

potetisensei commented on test/util/random.cpp:
```diff
@@ -73,16 +73,18 @@ BOOST_AUTO_TEST_CASE(TestWalkerDiscreteDistributionSimple) {
-    BOOST_CHECK(base[i] * 0.95 < res_p[i] && base[i] * 1.05 > res_p[i]);
+    const double p = base[i] / 1.0;  // w_i / sum(w)
+    const double E = iter * p;       // expectation
+    const double V = iter * (1 - p); // variance
```
This should be `iter * p * (1 - p)`, right?

potetisensei commented on test/util/random.cpp:
```diff
@@ -73,16 +73,18 @@ BOOST_AUTO_TEST_CASE(TestWalkerDiscreteDistributionSimple) {
+    res[index]++;
   }
 
+  constexpr double Z = 1.96; // alpha=0.05, Z_{0.025}
```
I understand that, in this part, the statistical test against binomial distributions is performed `len` times.
Probably it would be helpful if you leave a comment about it.
Also, as a CI test, I prefer much smaller p-values because there is a 5% chance of failure.

potetisensei commented on test/util/random.cpp:
```diff
@@ -95,12 +97,17 @@ BOOST_AUTO_TEST_CASE(TestWalkerDiscreteDistributionBiased) {
+    BOOST_CHECK(index < w1.size());
+    res1[index]++;
   }
+  constexpr double Z = 1.96;
```
The same as TestWalkerDiscreteDistributionSimple.

potetisensei commented on test/util/random.cpp:
```diff
@@ -95,12 +97,17 @@ BOOST_AUTO_TEST_CASE(TestWalkerDiscreteDistributionBiased) {
-    BOOST_CHECK(pa*0.9 <= ps && ps <= pa*1.1);
+    const double p = w1[i] / sum;
+    const double E = iter * p;
+    const double V = iter * (1 - p);
```
The same as TestWalkerDiscreteDistributionSimple.

potetisensei reviewed (changes requested):
First of all, I'm really sorry for my late re-review.
This commit is now almost perfect. But I found two concerns, so would you check them?

ptr-yudai committed:
:rotating_light: Fix test and add test for statistical test

ptr-yudai commented:
Fixed.
I changed Z score so that the test fails only with 0.1% probability:
https://github.com/fuzzuf/fuzzuf/blob/ac883de1d9a868f1efc4fcba0087d102effcefd3/test/util/random.cpp#L60

Also a test for statistical test is added:
https://github.com/fuzzuf/fuzzuf/blob/ac883de1d9a868f1efc4fcba0087d102effcefd3/test/util/random.cpp#L96-L98

I confirmed no error happened when I run the whole tests hundreds of times.

potetisensei reviewed (approved):
Great.
Maybe statisticians would get angry to see we use multiple binomial tests, not one chi-square test, which results in multiple comparisons problem.
But considering the used p-value, number of choices, and number of iterations, that concern would be negligible. If a problem occurs, we can fix it later.
Therefore I approve this commit.
